### DISCUSSION TITLE
Would like to help with OpcUa

### DISCUSSION THREAD

**Discussion by Pyroarti** [pyscada/PyScada#141] [General]
Hello.

I'm interested in contributing to this repository, particularly in enhancing the testing related to OpcUa. I've noticed there's a need for more comprehensive testing in this area. Could you share which aspects have already been tested? I have access to multiple PLCs operating an OpcUa server, and my PC also runs a server, providing me with hardware for testing. I would appreciate more details on the specific testing requirements and how I can best support this effort.

**Comment by clavay** [comment]
Hello, 

I worked with OPC device, I don't remember the name but I can search it in 2 weeks.
I was able to read from the device but I don't remember if I tried to write to the device.

If you want to improve the [pyscada-opcua](https://github.com/pyscada/PyScada-OPCUA) plugin, first you need to know how to communicate with your device.

The plugin use the [opcua-asyncio](https://github.com/FreeOpcUa/opcua-asyncio) library.

Look at the `_read_data` and `_write_data` functions I implemented in this [file](https://github.com/pyscada/PyScada-OPCUA/blob/main/pyscada/opcua/devices/__init__.py) to understand the logic.

If you are not able to communicate through the pyscada plugin, try to communicate directly using the librairy and send me the script which is working.

Let me know if you need help.

**Reply by Pyroarti** [reply]
Hello

Sounds good, i have a Phoenix PLCNext and a Siemens PLC ill use to test it when i have some spare time.

**Comment by clavay** [comment]
I updated the PyScada-OPCUA plugin and made it work with the [server example](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/server-minimal.py).

Please open an issue in the PyScada-OPCUA repository if needed.
